lib/types/zpool.nix

```diff
@@ -181,18 +210,31 @@ in
               else
                 entries=()
                 ${lib.optionalString (hasTopology && topology.vdev != null)
-                    (lib.concatMapStrings formatVdev topology.vdev)}
-                ${lib.optionalString (hasTopology && topology.special != null)
-                    (formatOutput "special ${topology.special.mode}" topology.special.members)}
+                    (formatVdevList "" topology.vdev)}
+                ${lib.optionalString (hasTopology && topology.spare != [])
+                    (formatOutput "spare" "" topology.spare)}
+                ${lib.optionalString (hasTopology && topology.log != [])
+                    (formatVdevList "log" topology.log)}
+                ${lib.optionalString (hasTopology && topology.dedup != [])
+                    (formatVdevList "dedup" topology.dedup)}
+                ${lib.optionalString (hasTopology && topology.special != null && topology.special != [])
+                    (formatVdevList "special" (lib.lists.toList topology.special))}
                 ${lib.optionalString (hasTopology && topology.cache != [])
-                    (formatOutput "cache" topology.cache)}
+                    (formatOutput "cache" "" topology.cache)}
                 all_devices=()
+                last_type=
                 for line in "''${entries[@]}"; do
-                  # lineformat is mode=device1 device2 device3
-                  mode=''${line%%=*}
-                  devs=''${line#*=}
+                  # lineformat is type mode=device1 device2 device3
+                  mode="''${line%%=*}"
+                  type="''${mode%% *}"
+                  mode="''${mode#"$type "}"
+                  devs="''${line#*=}"
                   IFS=' ' read -r -a devices <<< "$devs"
                   all_devices+=("''${devices[@]}")
+                  if ! [ "$type" = "$last_type" ]; then
+                    topology+=" $type"
+                    last_type="$type"
```

Comment:
What is last_type used for?

Reply:
It's there to ensure the type's keyword isn't repeated. If you want for example two log mirror vdevs you need to do 

```shell
zpool create -f <name> /dev/sda log mirror /dev/sdb /dev/sdc mirror /dev/sdd /dev/sde
```

and **not**

```shell
zpool create -f <name> /dev/sda log mirror /dev/sdb /dev/sdc log mirror /dev/sdd /dev/sde
```

Reply:
It is actually, all the special types that support actual vdevs (`log`, `dedup`, and `special`) have two in `zfs-with-vdevs`, and those would all run into the issue that this code avoids. In that example the vdevs are it's a single disk + a mirror, but the principle still applies.

```nix
{
  log = [
    {
      mode = "mirror";
      members = [ "log1" "log2" ];
    }
    {
      members = [ "log3" ];
    }
  ];
}
```

would result in

```shell
zpool create ... log /dev/log3 log mirror /dev/log1 /dev/log2 ...
```

without this logic (which results in the error `invalid vdev specification: 'log' can be specified only once`), and results in

```shell
zpool create ... log /dev/log3 mirror /dev/log1 /dev/log2 ...
```

with it.

Reply:
Ahh you're absolutely right!